### PR TITLE
feat: support whitelabel

### PR DESCRIPTION
**Pull request by wa0x6e**
### Summary

Closes: https://github.com/snapshot-labs/workflow/issues/142

This PR shows a special version of the website (white label version*), when accessed through a valid custom domain).

A white label is a stripped down version of the main website, with all Snapshot branding removed, with the site appearing like something owned by the DAO. No inside links to snapshot should remain, everything on the website should only concern the DAO.

A custom domain is a domain name (e.g. `vote.balancer.fi`) associated to a space id (`balancer.eth`), saved and served through a list managed by ourself. Currently, we used the hub, through the `Space` object to fetch a space, given a domain, so this feature is offchain spaces only.

### Main features

- Some components edited to be hidden on white label, e.g. sidebar, notifications, follow list, follow button, etc ....
- white label have its own set of routes (notably, the removal of `/network:id` prefix for space routes)
- white label do not have non-dao related pages (no global user pages, my- and site- routes, etc ...)
- the space breadcrumb should appear on all pages
- Added a new splash screen, to be shown while resolving the white label, or show an error when white label failed to resolve
- Removed the `space` key in most of the routes nested under `space-`, as this key will be inherited automatically when nested and when needed. Skipping this hardcoding allow routes to be compatible with white label routes, which do not have this params

### Changes summary 

- Most of the non-directly related changes have been offloaded into their own PR, so this one can be reviewed more easily
- Added a new `useWhiteLabel` composable
- Split routes into multiple files, index.ts is for logic, and the rest are for declaration.
- When using a custom domain, router will load a new catch all routes, and show a splash screen, while it's deciding whether the custom domain is associated to a space. If true, will load the white label routes and show the space, else load the main website routes and will show the main website.

### How to test

White label domain<->spaceID mapping list can be overridden locally by using the custom env variable `VITE_WHITE_LABEL_MAPPING`, formatted like `DOMAIN;SPACEID` (e.g. `VITE_WHITE_LABEL_MAPPING='vote.balancer.fi:balancer.eth'`)

With `VITE_WHITE_LABEL_MAPPING='127.0.0.1;balancer.eth'`, http://localhost:8080 will serve the snapshot website, whereas http://127.0.0.1:8080 will serve the balancer.eth white label version

#### Testing a custom domain associated to a space

1. Start the dev server with `VITE_WHITE_LABEL_MAPPING='127.0.0.1;balancer.eth yarn dev`
2. Go to http://localhost:3000
3. It should show the main website as before (immediate loading, no splash screen)
4. Go to http://127.0.0.1:8080
5. It should show the white label site for balancer.eth (will show a splash screen, but since on local, too fast to see it)

To test the splash screen, add a delay

```diff
diff --git a/apps/ui/src/composables/useWhiteLabel.ts b/apps/ui/src/composables/useWhiteLabel.ts
index 00a82383..0b6217d7 100644
--- a/apps/ui/src/composables/useWhiteLabel.ts
+++ b/apps/ui/src/composables/useWhiteLabel.ts
@@ -1,3 +1,4 @@
+import { sleep } from '@/helpers/utils';
 import { getNetwork } from '@/networks';
 import { useSpacesStore } from '@/stores/spaces';
 import { NetworkID } from '@/types';
@@ -29,6 +30,7 @@ async function getSpaceId(domain: string): Promise<string | null> {
   // e.g. VITE_WHITE_LABEL_MAPPING='127.0.0.1;snapshot.eth'
   const localMapping = import.meta.env.VITE_WHITE_LABEL_MAPPING;
   if (localMapping) {
+    await sleep(1000); // Add 1s delay to simulate network latency
     const [localDomain, localSpaceId] = localMapping.split(';');
     if (domain === localDomain) return localSpaceId;
   }
```

Loading http://127.0.0.1:8080 will now show the splash screen for 1s (emulating the delay for resolving the custom domain from the hub)

#### Testing a custom domain not associated to a space 

1. `yarn dev`
2. Go to http://127.0.0.1:8080
3. It should show the splash screen
4. It should load the main website after deciding that no space is associated to 127.0.0.1
5. Go to http://localhost:3000
6. It should show the main website as before (immediat loading, no splash **screen)**

#### Testing a custom domain not being able to resolve

1. Edit your code source and put an invalid url for the snapshot hub url
2. `yarn dev`
3. Go to http://127.0.0.1
4. It should show the splash screen
6. Once the hub fails to resolve, it should show the splash screen with an error message

### To-Do

- [x] App should respond to any domains thrown at it, in addition to localhost
- [x] Hide sidebar
- [x] Custom routes only related to /space
- [x] Remove Follow space button
- [x] Remove link to user global profile
- [x] Remove verified/turbo badge
- [x] My Profile should redirect to user space profile
- [x] Handle routes redirection on white label load
- [x] Check all "share" links
- [x] Check all Page title
- [x] Check for extrapolated url in router (.push(`xxx/xxx`)
- [x] Disable intercom on whitelabel
- [x] Fix duplicate calls of `fetchSpace`
- [x] Editor page crashing with resolver error
- [x] Set correct page title (current issue where page title is overwritten by space name)

### CODE REVIEW

**Comment by wa0x6e** [comment]
@bonustrack PR ready to test on the feature and UI/UX level

Code may need some additional tweakings, but UI should meet the specs, let's start the review as it may take time, and requires more changes

**Comment by bonustrack** [comment]
**Without whitelabel**

- When I resize my browser from small to large the left nav is cut
image

- Vote page have horizontal scroll:
image

**With whitelabel**
- For the loading, can you remove the loading icon, reduce the logo by ~20% use border color and `animate-pulse` instead?

File: apps/ui/index.html [hunk 1]
```diff
@@ -3,7 +3,7 @@
   <head>
     <meta charset="UTF-8" />
     <meta name="viewport" content="width=device-width, initial-scale=1.0" />
-    <title>Snapshot</title>
+    <title></title>
```
Comment (ChaituVR) [inline apps/ui/index.html:6]: I think we should still keep Snapshot here by default. (can be changed later) At least for some of the crawlers

File: apps/ui/vite.config.ts [hunk 2]
```diff
@@ -15,6 +15,9 @@ const target = ['esnext'];
 
 export default defineConfig({
   base: ELECTRON ? path.resolve(__dirname, './dist') : undefined,
+  server: {
+    host: '127.0.0.1'
+  },
```
Comment (ChaituVR) [inline apps/ui/vite.config.ts:20]: we will not have localhost here? 
image

File: apps/ui/src/composables/useWhiteLabel.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+async function getSpaceId(domain: string): Promise<string | null> {
+  // Resolve white label domain locally if mapping is provided
+  // for easier local testing
+  // e.g. VITE_WHITE_LABEL_MAPPING='127.0.0.1;s:snapshot.eth'
+  const localMapping = import.meta.env.VITE_WHITE_LABEL_MAPPING;
+  if (localMapping) {
+    const [localDomain, localSpaceId] = localMapping.split(';');
+    if (domain === localDomain) return localSpaceId;
+  }
```
Comment (ChaituVR) [inline apps/ui/src/composables/useWhiteLabel.ts:29]: Im not able to get it working with 

`VITE_WHITE_LABEL_MAPPING='127.0.0.1;s:thanku.eth` 

when I go to http://127.0.0.1:8081/# it still show normal app
maybe I miss some step?

Comment (wa0x6e) [reply apps/ui/src/composables/useWhiteLabel.ts:29]: What's your full command ?

Comment (ChaituVR) [reply apps/ui/src/composables/useWhiteLabel.ts:29]: i have added it .env and then `yarn dev`

Comment (wa0x6e) [reply apps/ui/src/composables/useWhiteLabel.ts:29]: try `VITE_WHITE_LABEL_MAPPING='127.0.0.1;s:snapshot.eth' yarn dev`

**Comment by wa0x6e** [comment]
Last minute update in the test env var, network id has been removed, now it's just `domain:spaceId`, e.g. `VITE_WHITE_LABEL_MAPPING='127.0.0.1;balancer.eth'`

**Comment by bonustrack** [comment]
> Last minute update in the test env var, network id has been removed, now it's just `domain:spaceId`, e.g. `VITE_WHITE_LABEL_MAPPING='127.0.0.1;balancer.eth'`

Any reason for that? Would that work with SX spaces too?

**Comment by wa0x6e** [comment]
Before it was working with sx, but with this change, not anymore.

Will bring that back if we add support for sx, but for now, it's simpler, and we only support offchain since we're resolving domain name with the hub

**Review by bonustrack — APPROVED** [review]
Tried all the ways and it works well! tACK

File: apps/ui/src/stores/spaces.ts [hunk 4]
```diff
@@ -34,10 +35,10 @@ export const useSpacesStore = defineStore('spaces', () => {
     };
   }
 
-  async function fetchSpaces(spaceIds: string[]) {
-    const spaces = await getSpaces({
-      id_in: spaceIds
-    });
+  async function fetchSpaces(input: string[] | SpacesFilter) {
+    const spaces = await getSpaces(
+      Array.isArray(input) ? { id_in: input } : input
```
Comment (ChaituVR) [inline apps/ui/src/stores/spaces.ts:40]: But we are not passing anything other than array any where? we need this condition?

File: apps/ui/src/App.vue [hunk 5]
```diff
@@ -1,14 +1,53 @@
...
+
+onMounted(() => {
+  initWhiteLabel();
+});
 </script>
 
 <template>
-  <LayoutSite v-if="isSiteRoute" />
+  <LayoutSplashScreen
+    v-if="routeName == 'splash-screen' || !whiteLabelResolved"
```
Comment (ChaituVR) [inline apps/ui/src/App.vue:49]: We need this condition `routeName == 'splash-screen'` ?

**Review by ChaituVR — APPROVED** [review]
tAck

**Comment by Sekhmet** [comment]
Didn't get around to testing it today, will review it tomorrow.

File: apps/ui/src/composables/useWhiteLabel.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
+import { getNetwork } from '@/networks';
+import { useSpacesStore } from '@/stores/spaces';
+import { Space } from '@/types';
+
+const NETWORK = 's';
```
Comment (Sekhmet) [inline apps/ui/src/composables/useWhiteLabel.ts:5]: Should we use this instead?
https://github.com/snapshot-labs/sx-monorepo/blob/4bfba1c11252bdbab241677ca7734ec7f0828dbb/apps/ui/src/networks/index.ts#L44-L45

File: apps/ui/src/components/Layout/SplashScreen.vue [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+const { resolved: whiteLabelResolved, failed } = useWhiteLabel();
+const { setAppName } = useApp();
+const router = useRouter();
+
+function handleReloadClick() {
+  window.location.reload();
+}
+
+watch(
+  () => whiteLabelResolved.value,
```
Comment (Sekhmet) [inline apps/ui/src/components/Layout/SplashScreen.vue:11]:
```suggestion
  whiteLabelResolved,
```

File: apps/ui/src/components/Layout/App.vue [hunk 8]
```diff
@@ -94,19 +109,46 @@ watch(scrollDisabled, val => {
...
-      (direction.value === 'left' && uiStore.sideMenuOpen))
+    (direction.value === 'right' && !uiStore.sideMenuOpen) ||
+    (direction.value === 'left' && uiStore.sideMenuOpen)
   ) {
     uiStore.toggleSidebar();
   }
 });
 
+watch(
+  () => isWhiteLabel.value,
```
Comment (Sekhmet) [inline apps/ui/src/components/Layout/App.vue:130]:
```suggestion
  isWhiteLabel,
```

File: apps/ui/src/views/Space.vue [hunk 9]
```diff
@@ -33,7 +38,9 @@ watchEffect(() => {
 });
 
 watchEffect(() => {
-  if (!space.value) return setFavicon(null);
+  if (!space.value) {
+    return setFavicon(null);
+  }
```
Comment (Sekhmet) [inline apps/ui/src/views/Space.vue:43]: Maybe we should quite early like in `unOnmounted` as this is already handled for whitelabel in LayoutApp.

File: apps/ui/src/components/Layout/App.vue [hunk 10]
```diff
@@ -94,19 +109,46 @@ watch(scrollDisabled, val => {
...
+    (direction.value === 'right' && !uiStore.sideMenuOpen) ||
+    (direction.value === 'left' && uiStore.sideMenuOpen)
   ) {
     uiStore.toggleSidebar();
   }
 });
 
+watch(
+  isWhiteLabel,
+  async isWhiteLabel => {
```
Comment (Sekhmet) [inline apps/ui/src/components/Layout/App.vue:131]:
```suggestion
  isWhiteLabel => {
```

File: apps/ui/src/App.vue [hunk 11]
```diff
@@ -1,14 +1,51 @@
...
+  const routes = isWhiteLabel.value ? whiteLabelRoutes : defaultRoutes;
+
+  routes.forEach(route => router.addRoute(route));
+  router.removeRoute('splash-screen');
+}
+
+watchEffect(() => setTitle(app.value.app_name));
+
+watch(
+  () => whiteLabelResolved.value,
```
Comment (Sekhmet) [inline apps/ui/src/App.vue:30]:
```suggestion
  whiteLabelResolved,
```

**Review by Sekhmet — APPROVED** [review]
tACK

**Comment by Sekhmet** [comment]
What is the plan regarding merging it, do we have any plans for it or we can just merge it and set it up later?

**Comment by wa0x6e** [comment]
I would prefer to merge asap, to avoid merge conflicts

**Comment by Sekhmet** [comment]
Let's merge this then, we can revert if issues show up and patching is too hard.

**Pull request merged**